R/read-gs.R

```diff
@@ -15,11 +16,200 @@ read_gs <- function(path, verbose = FALSE){
+#'@importFrom readr read_csv
+#'@export
+read_files <- function(grades_path, 
+                       other_file_paths = list(), 
+                       source = "auto"){
+  #first read in all of the files
+  
+  #begininng with the grades csv
+  
+  tryCatch({
+    grades <- readr::read_csv(grades_path, trim_ws = FALSE)
+    },
+    error = function(e){
+      stop(
+      paste0("Error when attempting to read grade csv file. 
+           Check that the filepath is correct. Error: ",
+             e)
+      )
+    }
+  )
+  
+  #now read in other supplied data 
+ for (data_type in names(other_file_paths)){
+   warning(paste0(data_type, 
+                  " is not currently supported as an extra file source."))
+ }
+  
+  # now determine how to read in data
+  
+  auto_determine <- determine_grade_source(grades)
+  
+  if (source == "auto"){
+    #fill in later
+    
+    if (auto_determine == "Gradescope"){
+      attr(grades, "source") <- "Gradescope"
+      
+      return(check_data_format(grades))
+    } else if (auto_determine == "Canvas"){
+      #the read_canvas_grades does all of the lifting
+      return(read_canvas_grades(grades))
+    } else {
+      stop("Data source is unrecognized.")
+    }
+    
+  } else if (source == "Gradescope"){
+    if (source != auto_determine){
+      warning(paste0(
+        "Grade data will be processed as ", source, 
+        " data due to value of source parameter. ",
+        "However, the data appears to be from source ",
+        auto_determine, "."
+      ))
+    }
+    
+    attr(grades, "source") <- "Gradescope"
+    
+    return(check_data_format(grades))
+      
+  } else if (source == "Canvas"){
+    if (source != auto_determine){
+      warning(paste0(
+        "Grade data will be processed as ", source, 
+        " data due to value of source parameter. ",
+        "However, the data appears to be from source ",
+        auto_determine, "."
+      ))
+    }
+    #the read_canvas_grades does all of the lifting
+    return(read_canvas_grades(grades))
+    
+  } else {
+    stop(paste0(
+      "Specified source type ", source,
+      " is not currently supported."
+    ))
+  }
+  
+  
+ 
+}
+
+#'Convert Canvas grade dataframe into standard format
+#'
+#'Takes in a dataframe uploaded from Canvas. 
+#'Returns a dataframe in our standardized format (ie like Gradescope).
+#'
+#'
+#'@importFrom stringr str_extract str_match
+#'@importFrom dplyr filter select slice rename_with rename left_join
+#'@importFrom purrr discard
+#'@keywords internal
+read_canvas_grades <- function(grades){
+  #first determine which columns are assignments
+  
+  #canvas categories will not be kept
+  
+  assignments <- stringr::str_extract(names(grades),
+                                      ".+\\s*\\(\\d+\\)") |>
+                purrr::discard(is.na)
+  
+ 
+  #now reconfigure the points possible info
+  max_points <- dplyr::filter(grades, Student == "    Points Possible") |>
```

Comment:
Woah, is `"    Points Possible"` truly the string they use here?! (with that arbitrary number of spaces)

Reply:
Yeah! I think they are aiming for nice aesthetics when looking at the csv in excel.